crates/valence_new/src/inventory.rs

```diff
@@ -104,58 +104,71 @@ impl Inventory {
+    fn slot_slice(&self) -> &[Option<ItemStack>] {
         self.slots.as_ref()
     }
 }
 
 /// Send updates for each client's player inventory.
-pub(crate) fn update_player_inventories(mut query: Query<(&mut Inventory, &mut Client)>) {
+pub(crate) fn update_player_inventories(
+    mut query: Query<(&mut Inventory, &mut Client), Without<OpenInventory>>,
+) {
     for (mut inventory, mut client) in query.iter_mut() {
         if inventory.kind != InventoryKind::Player {
             warn!("Inventory on client entity is not a player inventory");
         }
 
         if inventory.modified != 0 {
-            inventory.state_id += 1;
-
             if inventory.modified == u64::MAX {
                 // Update the whole inventory.
+                client.inventory_state_id += 1;
                 let cursor_item = client.cursor_item.clone();
+                let state_id = client.inventory_state_id.0;
                 client.write_packet(&SetContainerContentEncode {
                     window_id: 0,
-                    state_id: VarInt(inventory.state_id.0),
+                    state_id: VarInt(state_id),
                     slots: inventory.slot_slice(),
                     carried_item: &cursor_item,
                 });
 
                 client.cursor_item_modified = false;
             } else {
-                // Update only the slots that were modified.
-                for (i, slot) in inventory.slots.iter().enumerate() {
-                    if (inventory.modified >> i) & 1 == 1 {
-                        client.write_packet(&SetContainerSlotEncode {
-                            window_id: 0,
-                            state_id: VarInt(inventory.state_id.0),
-                            slot_idx: i as i16,
-                            slot_data: slot.as_ref(),
-                        });
+                // send the modified slots
+                if inventory.modified & !client.inventory_slots_modified != 0 {
+                    client.inventory_state_id += 1;
+                    let state_id = client.inventory_state_id.0;
+                    for (i, slot) in inventory.slots.iter().enumerate() {
+                        if (inventory.modified >> i) & 1 == 1 {
+                            if (client.inventory_slots_modified >> i) & 1 == 1 {
+                                // This slot was modified by this client, so we don't need to
+                                // send it
+                                continue;
+                            }
+                            client.write_packet(&SetContainerSlotEncode {
+                                window_id: 0,
+                                state_id: VarInt(state_id),
+                                slot_idx: i as i16,
+                                slot_data: slot.as_ref(),
+                            });
+                        }
                     }
                 }
             }
 
             inventory.modified = 0;
```

Comment:
Modified bits for the player inventory are not being cleared while an inventory is open because of the `Without<OpenInventory>` filter. This bug manifests like this:
1. Open a chest.
2. Move item from chest into player inventory.
3. When the chest is closed, you will be sent a `SetContainerSlot`

Solution: clear modified bits of all inventories in a dedicated system that runs after everything else? Then you wouldn't need the code at the bottom of `update_open_inventories`.

Reply:
I'm unable to reproduce this anymore.

Reply:
Seems to be fixed. Maybe a misdiagnosis from me.